File: assets/icons/generate-icons.py
#!/usr/bin/env python3
"""Generate simple PNG icons for the Liero PWA."""
import struct
import zlib
import math
import os

def create_png(size, output_path):
    """Create a PNG with dark purple bg, green circle, white 'L'."""
    bg = (0x1a, 0x0a, 0x2e)
    green = (0x00, 0xcc, 0x00)
    white = (0xff, 0xff, 0xff)

    cx, cy = size / 2, size / 2
    radius = size * 0.30

    # L letter dimensions
    l_left = cx - size * 0.10
    l_stroke = size * 0.06
    l_top = cy - size * 0.16
    l_bottom = cy + size * 0.16
    l_right = cx + size * 0.12
    l_foot_top = l_bottom - l_stroke

    raw = bytearray()
    for y in range(size):
        raw.append(0)  # filter byte
        for x in range(size):
            dx, dy = x - cx, y - cy
            dist = math.sqrt(dx*dx + dy*dy)

            if dist <= radius:
                in_vert = (l_left <= x <= l_left + l_stroke and l_top <= y <= l_bottom)
                in_horiz = (l_foot_top <= y <= l_bottom and l_left <= x <= l_right)
                if in_vert or in_horiz:
                    r, g, b = white
                else:
                    r, g, b = green
            elif dist <= radius + 2:
                t = (dist - radius) / 2
                r = int(green[0] * (1-t) + bg[0] * t)
                g = int(green[1] * (1-t) + bg[1] * t)
                b = int(green[2] * (1-t) + bg[2] * t)
            else:
                r, g, b = bg

            raw.extend([r, g, b])

    compressed = zlib.compress(bytes(raw), 9)

    def make_chunk(chunk_type, data):
        chunk = chunk_type + data
        crc = zlib.crc32(chunk) & 0xffffffff
        return struct.pack('>I', len(data)) + chunk + struct.pack('>I', crc)

    ihdr_data = struct.pack('>IIBBBBB', size, size, 8, 2, 0, 0, 0)

    png = b'\x89PNG\r\n\x1a\n'
    png += make_chunk(b'IHDR', ihdr_data)
    png += make_chunk(b'IDAT', compressed)
    png += make_chunk(b'IEND', b'')

    with open(output_path, 'wb') as f:
        f.write(png)

    print(f"Created {output_path} ({len(png)} bytes, {size}x{size})")

script_dir = os.path.dirname(os.path.abspath(__file__))
create_png(192, os.path.join(script_dir, 'icon-192.png'))
create_png(512, os.path.join(script_dir, 'icon-512.png'))

File: assets/icons/generate-icons.js
const zlib = require('zlib');
const fs = require('fs');
const path = require('path');

function createPNG(size, outputPath) {
  // Colors
  const bgR = 0x1a, bgG = 0x0a, bgB = 0x2e; // #1a0a2e dark purple
  const circR = 0x00, circG = 0xcc, circB = 0x00; // #00cc00 green
  const txtR = 0xff, txtG = 0xff, txtB = 0xff; // white

  const cx = size / 2;
  const cy = size / 2;
  const circleRadius = size * 0.30;

  // Simple "L" shape parameters
  const lLeft = cx - size * 0.10;
  const lRight = cx + size * 0.12;
  const lTop = cy - size * 0.16;
  const lBottom = cy + size * 0.16;
  const lStroke = size * 0.06;
  const lFootTop = lBottom - lStroke;

  // Build raw pixel data (filter byte + RGB for each row)
  const rawData = Buffer.alloc(size * (1 + size * 3));

  for (let y = 0; y < size; y++) {
    const rowOffset = y * (1 + size * 3);
    rawData[rowOffset] = 0; // filter: None

    for (let x = 0; x < size; x++) {
      const pixOffset = rowOffset + 1 + x * 3;
      const dx = x - cx;
      const dy = y - cy;
      const dist = Math.sqrt(dx * dx + dy * dy);

      let r, g, b;

      if (dist <= circleRadius) {
        // Inside green circle
        r = circR; g = circG; b = circB;

        // Draw "L" letter on top
        const inVerticalBar = (x >= lLeft && x <= lLeft + lStroke && y >= lTop && y <= lBottom);
        const inHorizontalBar = (y >= lFootTop && y <= lBottom && x >= lLeft && x <= lRight);
        if (inVerticalBar || inHorizontalBar) {
          r = txtR; g = txtG; b = txtB;
        }
      } else if (dist <= circleRadius + 2) {
        // Slight anti-alias / border
        const t = (dist - circleRadius) / 2;
        r = Math.round(circR * (1 - t) + bgR * t);
        g = Math.round(circG * (1 - t) + bgG * t);
        b = Math.round(circB * (1 - t) + bgB * t);
      } else {
        // Background
        r = bgR; g = bgG; b = bgB;
      }

      rawData[pixOffset] = r;
      rawData[pixOffset + 1] = g;
      rawData[pixOffset + 2] = b;
    }
  }

  // Compress pixel data
  const compressed = zlib.deflateSync(rawData, { level: 9 });

  // Build PNG file
  const chunks = [];

  // PNG Signature
  chunks.push(Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]));

  function writeChunk(type, data) {
    const typeBuffer = Buffer.from(type, 'ascii');
    const lenBuffer = Buffer.alloc(4);
    lenBuffer.writeUInt32BE(data.length, 0);

    const crcInput = Buffer.concat([typeBuffer, data]);
    const crc = crc32(crcInput);
    const crcBuffer = Buffer.alloc(4);
    crcBuffer.writeUInt32BE(crc >>> 0, 0);

    chunks.push(lenBuffer, typeBuffer, data, crcBuffer);
  }

  // CRC32 implementation
  function crc32(buf) {
    let table = [];
    for (let n = 0; n < 256; n++) {
      let c = n;
      for (let k = 0; k < 8; k++) {
        if (c & 1) c = 0xedb88320 ^ (c >>> 1);
        else c = c >>> 1;
      }
      table[n] = c;
    }
    let crc = 0xffffffff;
    for (let i = 0; i < buf.length; i++) {
      crc = table[(crc ^ buf[i]) & 0xff] ^ (crc >>> 8);
    }
    return (crc ^ 0xffffffff) >>> 0;
  }

  // IHDR chunk
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(size, 0);  // width
  ihdr.writeUInt32BE(size, 4);  // height
  ihdr[8] = 8;   // bit depth
  ihdr[9] = 2;   // color type: RGB
  ihdr[10] = 0;  // compression
  ihdr[11] = 0;  // filter
  ihdr[12] = 0;  // interlace
  writeChunk('IHDR', ihdr);

  // IDAT chunk
  writeChunk('IDAT', compressed);

  // IEND chunk
  writeChunk('IEND', Buffer.alloc(0));

  const png = Buffer.concat(chunks);
  fs.writeFileSync(outputPath, png);
  console.log(`Created ${outputPath} (${png.length} bytes, ${size}x${size})`);
}

const dir = path.dirname(process.argv[1]);
createPNG(192, path.join(dir, 'icon-192.png'));
createPNG(512, path.join(dir, 'icon-512.png'));
